src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -37,6 +37,30 @@ impl ScalarFunc {
         let i = try_opt!(self.children[0].eval_int(ctx, row));
         Ok(Some(Cow::Owned(format!("{:b}", i).into_bytes())))
     }
+
+    pub fn left<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string(ctx, row));
+        let i = try_opt!(self.children[1].eval_int(ctx, row));
+        if i <= 0 {
+            return Ok(Some(Cow::Owned(b"".to_vec())));
+        }
+        {
+            let s = str::from_utf8(&s)?;
+            if s.chars().count() > i as usize {
+                let t = s
+                    .chars()
+                    .into_iter()
+                    .map(|x| x.to_string())
```

Comment:
I think you can avoid some allocations by doing this:

```rust
            let s = str::from_utf8(&s)?;
            if s.chars().count() > i as usize {
                let t = s
                    .chars()
                    .into_iter();
                return Ok(Some(Cow::Owned(t.take(i as usize)
                                          .collect::<String>().into_bytes())));
            }
```

Reply:
ok